libs/api/domains/translations/src/lib/translations.service.ts

```diff
@@ -33,75 +34,67 @@ const errorHandler = (name: string) => {
 
 @Injectable()
 export class TranslationsService {
-  loadedNamespaces = new Map<
-    string,
-    { id: Locale; messages: TranslationsDict }[]
-  >()
-  fetching: Record<string, boolean> = {}
-
   constructor(private contentfulRepository: ContentfulRepository) {}
 
-  getTranslations = async (
-    namespaces?: string[],
-    lang?: Locale,
-  ): Promise<TranslationsDict> => {
-    const locale = locales.find((l) => l.code === localeMap[lang ?? 'is']) as {
-      code: string
-      fallbackCode: string | null
-    }
-
-    if (!namespaces || !lang) {
-      throw new Error('No namespaces or lang defined.')
-    }
-
-    const result = await this.contentfulRepository
-      .getLocalizedEntries<NamespaceFields>('*', {
-        ['content_type']: 'namespace',
-        select: 'fields.strings',
-        'fields.namespace[in]': namespaces.join(','),
-      })
-      .catch(errorHandler('getNamespace'))
-
-    return result.items.reduce((acc: TranslationsDict, cur) => {
-      const strings = cur.fields.strings
-
-      return {
-        ...acc,
-        ...mergeWith(
-          {},
-          locale.fallbackCode ? strings?.[locale.fallbackCode] : {},
-          strings?.[locale.code],
-          (o, s) => (isEmpty(s) ? o : s),
-        ),
-      }
-    }, {})
-  }
-
-  fetchNamespaces = async (namespaces: string[]) => {
-    await Promise.all(
-      namespaces.map(async (namespace) => {
-        if (this.fetching?.[namespace]) {
-          return
+  getOrRequest = memoize(
+    async (namespace: string) =>
+      await this.contentfulRepository
+        .getLocalizedEntries<NamespaceFields>('*', {
+          ['content_type']: 'namespace',
+          select: 'fields.strings',
+          'fields.namespace[in]': namespace,
+        })
+        .catch(errorHandler('getNamespace')),
+    { maxAge: MAX_AGE, preFetch: true },
+  )
+
+  getMessages = async (entries: EntryCollection<NamespaceFields>[]) => {
+    let messages: Record<string, TranslationsDict> = {}
+
+    for (const namespace of entries) {
+      for (const item of namespace.items) {
+        const strings = item.fields.strings
+
+        if (!strings) {
+          continue
         }
 
-        this.fetching[namespace] = true
-
-        if (this.loadedNamespaces.has(namespace)) {
-          logger.info(`${namespace} already exists.`)
-
-          return
+        for (const contentfulLocale of Object.keys(strings)) {
+          const locale = locales.find((item) => item.code === contentfulLocale)!
+            .locale as Locale
+
+          for (const key of Object.keys(strings[contentfulLocale])) {
+            const message =
+              contentfulLocale === DEFAULT_LOCALE
+                ? strings?.[contentfulLocale]?.[key]
+                : strings?.[contentfulLocale]?.[key] !== ''
```

Comment:
Is this check why it was using `mergeWith` before? 

I'd like to optimize this function quite a bit, to not process unused locales, and not create Namespaces * Locales * Keys growing objects.

I'd suggest moving the fallback logic behind the memoized function, so it returns an object that looks like this: `{ is: allIcelandicTranslations, en: allEnglishTranslationsWithFallbacks }`. Then we can easily loop through the requested namespaces and merge the requested locale into a final object.

I generally avoid doing `{ ...obj, [key]: value }` many times in the hot path (code that runs a lot) and modify a mutable object instead. But this is less critical if we move it behind memoizee.